packages/java-parser/src/comments.js

```diff
@@ -0,0 +1,156 @@
+  const attachComments = [...comments];
+
+  // edge case: when the file contains only one token (;/if no token then EOF)
+  if (tokens.length === 1) {
+    attachComments.forEach(comment => {
+      if (comment.endOffset < tokens[0].startOffset) {
+        if (!tokens[0].leadingComments) {
+          tokens[0].leadingComments = [];
+        }
+        tokens[0].leadingComments.push(comment);
+      } else {
+        if (!tokens[0].trailingComments) {
+          tokens[0].trailingComments = [];
+        }
+        tokens[0].trailingComments.push(comment);
+      }
+    });
+    return tokens;
+  }
+
+  // edge case: when the file start with comments, it attaches as leadingComments to the first token
+  const firstToken = tokens[0];
+  const headComments = [];
+  while (
+    attachComments.length > 0 &&
+    attachComments[0].endOffset < firstToken.startOffset
+  ) {
+    headComments.push(attachComments[0]);
+    attachComments.splice(0, 1);
+  }
+
+  if (headComments.length > 0) {
+    firstToken.leadingComments = headComments;
+  }
+
+  // edge case: when the file end with comments, it attaches as trailingComments to the last token
+  const lastToken = tokens[tokens.length - 1];
+  const tailComments = [];
+  while (
+    attachComments.length > 0 &&
+    attachComments[attachComments.length - 1].startOffset > lastToken.endOffset
+  ) {
+    tailComments.push(attachComments[attachComments.length - 1]);
+    attachComments.splice(attachComments.length - 1, 1);
+  }
+
+  if (tailComments.length > 0) {
+    lastToken.trailingComments = tailComments.reverse();
+  }
+
+  let currentToken = 0;
+  attachComments.forEach(element => {
+    // find the correct position to place the comment
+    while (
+      !(
+        element.startOffset > tokens[currentToken].endOffset &&
+        element.endOffset < tokens[currentToken + 1].startOffset
+      )
+    ) {
+      currentToken++;
+    }
+
+    // attach comment to the closest token
+    if (
+      element.startOffset - tokens[currentToken].endOffset <
+      tokens[currentToken + 1].startOffset - element.endOffset
+    ) {
+      if (!tokens[currentToken].hasOwnProperty("trailingComments")) {
+        tokens[currentToken].trailingComments = [];
+      }
+      tokens[currentToken].trailingComments.push(element);
+    } else {
+      if (!tokens[currentToken + 1].hasOwnProperty("leadingComments")) {
+        tokens[currentToken + 1].leadingComments = [];
+      }
+      tokens[currentToken + 1].leadingComments.push(element);
+    }
+  });
+
+  return tokens;
+}
+
+function findUpperBoundToken(tokens, comment) {
+  let diff;
+  let i;
+  let current;
+
+  let len = tokens.length;
+  i = 0;
+
+  while (len) {
+    diff = len >>> 1;
+    current = i + diff;
+    if (tokens[current].startOffset > comment.startOffset) {
+      len = diff;
+    } else {
+      i = current + 1;
+      len -= diff + 1;
+    }
+  }
+  return i;
+}
+
+function extendCommentRange(tokens, comments) {
```

Comment:
this needs docs i dont understand what is happening here.

Reply:
The offsets on nodes are defined by the startOffset of the first token and the endOffset of the last token. However between a token and another one (comment in our case) there might be some spaces/tabs.
Thus, by extending the "comments" range to the endOffset of the previous token and the startOffset of the next token. We are able to know to which node we should attach as IMO would be much harder without it and will add more computing power if we do it when we traverse all the cst.

Reply:
I will comment and add docs on it. But I already planned to document the project.

Reply:
O.K I think I understand. You want to know the "real" token immediately following a comment and the "real" token just before a comment.

Would constructing a singular (sorted) combined Token Vector that combines both the "real" tokens and the comment tokens assist in this regard and/or be simpler?

Also does the logic below handle the edge case where there is a sequence of comment tokens?